### PR TITLE
Expose functional IR to capture_pre_autograd

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116383
* __->__ #115210
* #115188

cc @avikchaudhuri @gmagogsfm @zhxchen17 @angelayi @suo @ydwu4

### CODE REVIEW

File: torch/_export/__init__.py [hunk 1]
```diff
@@ -177,37 +178,46 @@ def capture_pre_autograd_graph(
...
+                range_constraints=range_constraints,
+                equality_constraints=equality_constraints,
+            )
+    else:
+        module = _export(f, args, kwargs, constraints=constraints, pre_dispatch=True, decomp_table=decomp_table).module()
+
+    def _train(self, mode: bool = True):
+        raise NotImplementedError("Calling train() is not supported yet.")
 
+<<<<<<< HEAD
```
Comment (bdhirsh) [inline torch/_export/__init__.py:220]: 🤔

File: torch/_export/__init__.py [hunk 2]
```diff
@@ -177,37 +178,46 @@ def capture_pre_autograd_graph(
         torch.ops.aten.native_batch_norm.default: torch.ops.aten.native_batch_norm.default.decompose,
     }
 
-    if kwargs is None:
-        kwargs = {}
+    if not _functional_pre_dispatch_IR:
```
Comment (bdhirsh) [inline torch/_export/__init__.py:181]: nit: I would probably invert the condition (to limit unnecessary indents)

File: test/export/test_export.py [hunk 3]
```diff
@@ -1412,13 +1412,15 @@ def forward(self, x):
         example_inputs = (torch.randn(1, 3, 3, 3),)
         m = CondBranchClassMethod()
         m.eval()
-        gm = capture_pre_autograd_graph(m, example_inputs)
+        # TODO (tmanlaibaatar) It doesn't work on aot_export yet
```
Comment (bdhirsh) [inline test/export/test_export.py:1415]: what doesn't work?

File: torch/_export/__init__.py [hunk 4]
```diff
@@ -262,6 +269,8 @@ def _export(
     constraints: Optional[List[Constraint]] = None,
     *,
     strict: bool = True,
+    decomp_table: Optional[Dict[str, Callable]] = None,
+    pre_dispatch: bool = False,
```
Comment (zhxchen17) [inline torch/_export/__init__.py:273]: @tugsbayasgalan is the plan to make this default to be True in the future?

Comment (tugsbayasgalan) [reply torch/_export/__init__.py:273]: Yep.

Comment (bdhirsh) [inline torch/_export/__init__.py:273]: Seems like these two arg aren't being used?

Comment (tugsbayasgalan) [reply torch/_export/__init__.py:273]: it is actually used here: https://github.com/pytorch/pytorch/pull/115210/files#diff-7c238fb036dbd2844e5ecdcb28e35f5cc3eb3a664f212f2d9a066e9c6ede7019R182

File: torch/_export/__init__.py [hunk 5]
```diff
@@ -177,51 +178,57 @@ def capture_pre_autograd_graph(
...
-        unlifted_m = _create_stateful_graph_module(
-            m,
-            range_constraints=range_constraints,
-            equality_constraints=equality_constraints,
-        )
-        unlifted_m.train = types.MethodType(_train, m)  # type: ignore[method-assign]
-        unlifted_m.eval = types.MethodType(_eval, m)  # type: ignore[method-assign]
-        return unlifted_m
+            flat_args, _ = pytree.tree_flatten((args, kwargs or {}))
+            range_constraints, equality_constraints = _process_constraints(m, 0, flat_args)
```
Comment (angelayi) [inline torch/_export/__init__.py:210]: why is this called twice?

File: torch/_export/__init__.py (hunk 4 above)
Comment (bdhirsh) [reply torch/_export/__init__.py:273]: nvm haha. I thought this was `torch/export/__init__.py`

File: torch/_export/__init__.py (hunk 5 above)
Comment (tugsbayasgalan) [reply torch/_export/__init__.py:210]: Good catch! Just some merge conflict shenanigans. Will delete

File: torch/_export/__init__.py (hunk 4 above)
Comment (angelayi) [reply torch/_export/__init__.py:273]: wait sorry, I'm still confused by these two arguments. They're not being used in _export though?

**Review by zhxchen17 — APPROVED** [review]

**Pull request closed without merging**